tianshou/data/buffer.py

```diff
@@ -359,6 +372,78 @@ def __getitem__(
             policy=self.get(index, "policy"),
         )
 
+    @classmethod
+    def _copy_to_hdf5(
+        cls, k: str, v: Union[np.ndarray, Batch], grp: h5py.Group
+    ) -> None:
+        if isinstance(v, np.ndarray):
+            grp.create_dataset(k, data=v)
+        elif isinstance(v, torch.Tensor):
+            grp.create_dataset(k, data=v.cpu().numpy())
+        elif isinstance(v, Batch):
+            subgrp = grp.create_group(k)
+            for bk, bv in v.__dict__.items():
+                cls._copy_to_hdf5(bk, bv, subgrp)
+
+    @classmethod
+    def _copy_from_hdf5(
+        cls, grp: h5py.Group, dst: Batch, device: str = "numpy"
+    ) -> None:
+        for k, v in grp.items():
+            if isinstance(v, h5py.Group):
+                if k not in dst.__dict__:
+                    dst.__dict__[k] = Batch()
+                cls._copy_from_hdf5(v, dst.__dict__[k])
+            elif isinstance(v, h5py.Dataset):
+                if k in dst.__dict__:
+                    if isinstance(dst.__dict__[k], np.ndarray):
+                        v.read_direct(dst.__dict__[k])
+                    elif isinstance(dst.__dict__[k], torch.Tensor):
+                        dst.__dict__[k] = torch.tensor(v, device=device)
+                    else:
+                        raise Exception("Cannot copy HDF5 dataset into object"
+                                        f"with type {type(dst.__dict__[k])}.")
+                else:
+                    if device == "numpy":
+                        dst.__dict__[k] = np.empty(v.shape, dtype=v.dtype)
+                        v.read_direct(dst.__dict__[k])
+                    else:
+                        dst.__dict__[k] = torch.tensor(v, device=device)
+
+    def save_hdf5(self, path: str) -> None:
+        """Save replay buffer to HDF5 file."""
+        with h5py.File(path, "w") as f:
+            for k, v in self.__dict__.items():
+                if k not in ["_meta", "_indices"]:
```

Comment:
I'm not comfortable with L417 and L430. `_meta` is fine but others are kind of tricky. It means we have to pay additional attention to these lines in future maintenance. Is there a better way of grp organization?

Reply:
I can see your point. Maybe an alternative way would be to introduce a method _get_state that returns a dictionary with all attributes that should be saved to a file when saving the replay buffer? At some point a distinction between attributes like _indices that do not have to be saved and those that should be saved has to be made.

Reply:
You can use `__getstate__` directly. It works.

Reply:
Can we build this feature only on top of `__getstate__` and `__setstate__`?

Reply:
Sounds good, however, HDF5 isn't really intended to be interchangeable with pickle. So I am not sure whether it is possible to write saving and loading code that won't break no matter what changes to the attributes are done. I will try to come up with something that is more general (like you suggested for group organization) and get back to you.

Reply:
Yes your are right. In the general case, hdf5 is not intenses to be interchangeable with pickle, but in this particular case, you are dealing with a class designed to store data. This in this particular setting, I think it should be possible to come up a common method to extract raw data from `batch` and restore them. Proabably it won't be possible to avoid having to implement a custom `from/to_hdf5` interface, but it should be reduced to bare minimum.